.github/workflows/rust.yml

```diff
@@ -39,6 +39,14 @@ on:
   workflow_dispatch:
 
 jobs:
+  # Check license header
+  license-header-check:
+    runs-on: ubuntu-20.04
+    name: Check License Header
+    steps:
+      - uses: actions/checkout@v4
+      - uses: korandoru/hawkeye@v5
```

Comment:
Is this an approved github action (I think we are restricted to running only approved actions in Apache repositories, but I haven't checked this particular one)

Note there is already a `RAT` check that is supposed to do this license checking, but clearly isn't catching some of this stuff

https://github.com/apache/datafusion/blob/ea2e7ab6885de734243dffc4642e2742206de5b9/.github/workflows/dev.yml#L27-L37

One benefit is the RAT check is the same check as used when creating a release candidate.

I think we should have only one way to check license headers, so if we go with this new action we should ensure it is also used when creating release candidates

Reply:
TIL, thanks for the info! I'll checkout RAT later

Reply:
After some searching, it seems like RAT is provided by a 3rd-party workflow and hasn't been moved into Apache org (https://issues.apache.org/jira/browse/RAT-337). Other projects that use RAT are not using its workflow directly but via some workaround like us via Python (e.g. `apache/arrow` is via pre-commit tool [here](https://github.com/apache/arrow/blob/07420b0c56066326bd409e9537ee3d43ab6b1a51/.pre-commit-config.yaml#L24-L31)). I'm not sure if RAT is a mature way in GitHub Action.

The [RAT check](https://github.com/apache/datafusion/blob/main/dev/release/run-rat.sh) in our repo hasn't been changed for 3 years 🤣 Maybe we need another tool for better check capability. I find OpenDAL also uses this workflow [here](https://github.com/apache/opendal/blob/e869a70a63c0e6ccaaeabe7678ec77ecd614f5ac/.github/workflows/ci_check.yml#L46-L54). So we can assume it's in the approved list?

Reply:
I am not sure about license / workflow. 

the run-rat.sh setup came from the https://github.com/apache/arrow repo when we broke arrow-rs and datafusion out into their own repos. I haven't followed any changes in the arrow repo since then

I don't personally have any preference about the mechanism of checking licenses as long as it is done.

I do think it is important that whatever CI uses is consistent with creating the release tarballs so we don't interfere with creating new releases:

https://github.com/apache/datafusion/blob/4838cfbf453f3c21d9c5a84f9577329dd78aa763/dev/release/create-tarball.sh#L120-L121

(perhaps this just means removing the explicit additional check for licenses when creating release tarballs, as they are redundant with the checks done on ci 🤔 )